### ISSUE TITLE
Memcached Ticket Registry Not Being Used

### ISSUE THREAD

**Issue by nattyddubbs** [apereo/cas#2912]
I am attempting to enable the memcached ticket registry on CAS 5.0.7 running in on Tomcat 8.5.5. I can see that CAS connects to the running memcached instance but then the default ticket registry is still used. 

Below is a snippet of logging from catalina.out when I start the server

```
2017-09-06 12:39:43.469 INFO net.spy.memcached.MemcachedConnection:  Added {QA sa=localhost/127.0.0.1:32769, #Rops=0, #Wops=0, #iq=0, topRop=null, topWop=null, toWrite=0, interested=0} to connect queue
2017-09-06 12:39:43,894 WARN [org.apereo.cas.services.InMemoryServiceRegistryDaoImpl] - <Runtime memory is used as the persistence storage for retrieving and persisting service definitions. Changes that are made to service definitions during runtime WILL be LOST upon container restarts.>
2017-09-06 12:39:43,922 INFO [org.apereo.cas.services.DefaultServicesManagerImpl] - <Loaded 0 services from InMemoryServiceRegistryDaoImpl.>
2017-09-06 12:39:49,342 INFO [org.apereo.cas.services.DefaultServicesManagerImpl] - <Loaded 4 services from InMemoryServiceRegistryDaoImpl.>
2017-09-06 12:39:56,497 INFO [org.apereo.cas.configuration.CasConfigurationRebinder] - <Reloaded CAS configuration cas-org.apereo.cas.configuration.CasConfigurationProperties>
2017-09-06 12:39:57,977 INFO [org.apereo.cas.configuration.support.Beans] - <Ticket registry encryption/signing is turned off. This MAY NOT be safe in a clustered production environment. Consider using other choices to handle encryption, signing and verification of ticket registry tickets, and verify the chosen ticket registry does support this behavior.>
2017-09-06 12:39:57,982 WARN [org.apereo.cas.ticket.registry.DefaultTicketRegistry] - <Runtime memory is used as the persistence storage for retrieving and managing tickets. Tickets that are issued during runtime will be LOST upon container restarts. This MAY impact SSO functionality.>
```

Below is my application.properties
```
cas.ticket.registry.memcached.servers=localhost:32769
cas.ticket.registry.memcached.locatorType=ARRAY_MOD
cas.ticket.registry.memcached.failureMode=Redistribute
cas.ticket.registry.memcached.hashAlgorithm=FNV1_64_HASH
cas.ticket.registry.memcached.crypto.signing.key=<secret>
cas.ticket.registry.memcached.crypto.signing.keySize=512
cas.ticket.registry.memcached.crypto.encryption.key=<secret>
cas.ticket.registry.memcached.crypto.encryption.keySize=16
cas.ticket.registry.memcached.crypto.alg=AES
```

I added the following to my pom.xml in the cas-server-components directory
```XML
<dependency>
            <groupId>org.apereo.cas</groupId>
            <artifactId>cas-server-support-memcached-ticket-registry</artifactId>
            <version>${cas.version}</version>
        </dependency>
```

I'm sure that I am overlooking something. I appreciate any help that can be provided.

**Comment by nattyddubbs** [comment]
Possibly related: https://groups.google.com/d/msg/jasig-cas-user/CUjsQbaUvso/csGwHlerAQAJ

**Comment by rdev5** [comment]
Verify your memcached server port is correct (i.e. 11211). Failing that, you might try another version.

I have the same exact settings and dependency and it works in CAS 5.1.1-SNAPSHOT. Using Gradle here, though.

**Comment by nattyddubbs** [comment]
@rdev5 The upgrade from 5.0.x to 5.1.x is more of a long term goal than an immediate option.

**Comment by mmoayyed** [comment]
Do you see the same thing with 5.0.8?

**Comment by mmoayyed** [comment]
I believe this is simply a "bad log message" where in fact the memcached ticket registry is used assuming you have it in your config. The log message just shows up a bit too eagerly which is of course wrong and misleading. Can you confirm this theory?

**Comment by nattyddubbs** [comment]
@mmoayyed I will check right now.

**Comment by nattyddubbs** [comment]
@mmoayyed You are correct the memcache does have the TGT cached.

```
stats cachedump 14 100       
ITEM TGT-1-prKZpGuEw5IyFG96YLdNawgpADpfgZYK9woyv9XbJVRKuelm5f-10 [1384 b; 1504737987 s]
ITEM TGT-1-efzfTeeubrJkEMx21PYCdhXPhLZfpPM4kbzdXIebJSctq060Bo-10 [1384 b; 1504737704 s]
ITEM TGT-1-2hd4z3GgCMQuojvYgtTfhV9tpdoSsjIgmQgcEgawB69Mw4yUUN-10 [1472 b; 1504737485 s]
END
```

As you alluded to this is a "bad log message". I could have done a little more due diligence to find that the cache had the tickets. Looking at the `CasCoreTicketsConfiguration ` it seems that the message is only logged when no other ticket registry bean has been created. Could this just be an order of initialization logging problem?

**Comment by mmoayyed** [comment]
> As you alluded to this is a "bad log message". I could have done a little more due diligence to find that the cache had the tickets. 

No worries. It sure is an annoying thing to figure out.

> Looking at the CasCoreTicketsConfiguration it seems that the message is only logged when no other ticket registry bean has been created. Could this just be an order of initialization logging problem?

Yes definitely. More specifically, I think it is somewhat caused by how `@ConditionalOnMissingBean` works where eventually the correct bean (read registry) is initialized to do its job yet the initialization process and specially the placement of the log message makes that rather unclear.

**Comment by mmoayyed** [comment]
More importantly, two things:

- It is in your best interest to upgrade to 5.1. In about two months, 5.0.x is going to transition into SPM mode:
https://apereo.github.io/cas/developer/Maintenance-Policy.html#eol-schedule

- It appears that this problem is fixed in 5.0.9 where you can exercise the latest snapshot to see the fix in action.
